### ISSUE TITLE
Inconsistent behaviour & syntax highlighting when creating lists

### ISSUE THREAD

**Issue by sjmduncan** [gsantner/markor#11]
#### General information

* **Device:**  Moto G 4G
* **Android Version:**  Android 7.1 (LineageOS 14.1)
* **App version:** v0.1.1

#### Steps to reproduce

1. Create a note
2. Add a list item (* + - or 1. ) 
3. Press enter at the end of the line

#### Expected result

1. Syntax-highlight the list marker
2. When creating a newline at the end of a list item, the next line should automatically start with the appropriate list marker, preserving indentation.

Syntax highlihghting doesn't work for the + or - markers. For the - marker every second item is highlighted with the header style. Both of those things work as expected in the markdown preview. 

Auto-inserting list markers only works for the * marker, only when editing existing notes, and only after you press enter at the end of a list item that already existed in that note when you loaded it.

**Comment by gsantner** [comment]
The markdown highlighting is not modified yet since forking. I am thinking of maybe swapping it out completly instead of fixing the broken implementation. If you find something that works good and can be added (open source lib) I am open to include and replace the current implemetnation with it.

You can send hints for a good replacement, or a PR to improve the highlighting, but I would prefer to use some maintained markdown-highlighter instead of some outdated custom soluton.

**Comment by sjmduncan** [comment]
There's [RxMarkdown](https://github.com/yydcdut/RxMarkdown) but I have no idea if it's any  good.

**Comment by gsantner** [comment]
Seems nice, but very heavy / pulls in too much dependencies

**Comment by sjmduncan** [comment]
For the moment I'll tidy up the existing one because I haven't found a decent replacement. It's not really *that* bad, mostly just some incomplete regexes.

It would be nice to have a renderer which is consistent with the syntax highlighter though. The current preview renderer has some issues of its own - no strikethrough, H1 with '===' but no H2 with '---', and sometimes weird list item spacing sometimes to name a few.

**Comment by gsantner** [comment]
If you want and can, you can also take a look into my SimpleMarkdownParser. That works really good in many apps yet (for displaying). I dont know how much would need to be done to make it highlighting too

**Comment by gsantner** [comment]
It's `net.gsantner.opoc.util.SimpleMarkdownParser`. If that can be expanded to highlighting too, I think it could be even more great. Basicially the infrastructure is there, it would just ned a custom "filter" (as you can see theres e.g. one for CHANGELOG especially)

**Comment by sjmduncan** [comment]
I already have a PR to fix the most obvious problems.

`SimpleMarkDownParser` could work, I'll look into that.

**Comment by gsantner** [comment]
k thx. The two default parser I think support most stuff of markdown yet. Except nested lists I think most is implemented. This is 100% dome by myself :D.

**Comment by sjmduncan** [comment]
@gsantner thoughts on/objections to using [commonmark-java](https://github.com/atlassian/commonmark-java)?

**Comment by gsantner** [comment]
The renderer is not our problem, that works fine so far. The problem is the highlighting editor, which does commonmark not provide

**Comment by sjmduncan** [comment]
@gsantner I asked because in lieu of a suitable and currently-maintained highlighting library the next best option is the current highlighter part of the editor (which is good at highlighting) with the same markdown parser that the renderer uses.

And I asked about the commonmark-java library because hoedown looks like it hasn't been actively developed for about 2 years, and it does a few weird things like sometimes wrapping the contents of list items in <p> tags and messing up the item spacing.

**Comment by gsantner** [comment]
another option is to let it be done via js, thats what octodroid is doing:

https://github.com/slapperwan/gh4a/blob/master/app/src/main/java/com/gh4a/widget/MarkdownPreviewWebView.java
